examples/icmp_response.rs

```diff
@@ -0,0 +1,266 @@
+    }
+
+    /* save source MAC address into variable */
+    let all_interfaces: Vec<NetworkInterface> = interfaces();
+    let is_exist: Option<&NetworkInterface> = all_interfaces
+        .iter()
+        .find(|element| element.name.as_str() == if_name.as_str());
+
+    let src_mac_address: MacAddr = match is_exist {
+        Some(target_interface) => {
+            let option: Option<MacAddr> = target_interface.mac;
+            if option.is_none() {
+                panic!("couldn't find source MAC address");
+            }
+
+            option.unwrap()
+        }
+        None => {
+            panic!("couldn't find source MAC address");
+        }
+    };
+
+    /* execute pv_open() */
+    let pv_open_option: Option<pv::Nic> = pv::pv_open(
+        &if_name,
+        chunk_size,
+        chunk_count,
+        rx_ring_size,
+        tx_ring_size,
+        filling_ring_size,
+        completion_ring_size,
+    );
+    let mut nic: pv::Nic;
+    if let Some(a) = pv_open_option {
+        nic = a;
+    } else {
+        panic!("nic allocation failed!");
+    }
+
+    /* initialize rx_batch_size and packet metadata */
+    let rx_batch_size: u32 = 64;
+    let mut packets: Vec<pv::Packet> = Vec::with_capacity(rx_batch_size as usize);
+
+    while !term.load(Ordering::Relaxed) {
+        let received: u32 = pv::pv_receive(&mut nic, &mut packets, rx_batch_size);
+
+        if received > 0 {
+            let processed: u32 =
+                process_packets(&mut nic, &mut packets, received, &src_mac_address);
+            let sent: u32 = pv::pv_send(&mut nic, &mut packets, processed);
+
+            if sent == 0 {
+                for i in (0..processed).rev() {
+                    pv::pv_free(&mut nic, &mut packets, i as usize);
+                }
+            }
+        }
+    }
+
+    pv::pv_close(nic);
+    println!("PV END");
+}
+
+fn process_packets(
+    nic: &mut pv::Nic,
+    packets: &mut Vec<pv::Packet>,
+    batch_size: u32,
+    src_mac_address: &MacAddr,
+) -> u32 {
+    let mut processed = 0;
+    for i in (0..batch_size as usize).rev() {
+        match chk_protocol(&packets[i]) {
+            "ARP" => {
+                if make_arp_response_packet(src_mac_address, &mut packets[i]).is_ok() {
+                    processed += 1;
+                } else {
+                    pv::pv_free(nic, packets, i);
+                }
+            }
+            "ICMP" => {
+                if make_icmp_response_packet(&mut packets[i]).is_ok() {
+                    processed += 1;
+                } else {
+                    pv::pv_free(nic, packets, i);
+                }
+            }
+            _ => {
+                pv::pv_free(nic, packets, i);
+            }
+        }
+    }
+    processed
+}
+
+// analyze what kind of given packet
+fn chk_protocol(packet: &pv::Packet) -> &str {
+    if is_icmp_req(packet) {
+        "ICMP"
+    } else if is_arp_req(packet) {
+        "ARP"
+    } else {
+        ""
+    }
+}
```

Comment:
Returning `enum` instead of `&str` seems better in this case and the function name `get_protocol()` looks clear because the function returns what type of protocol.

`enum ... {
    icmp,
    arp,
    unused,
 ...
}`